benches/nexmark/nexmark_dbsp_source.rs

```diff
@@ -0,0 +1,120 @@
+//! DBSP Source operator that reads from a Nexmark Generator.
+
+use crate::generator::{NexmarkGenerator, NextEvent};
+use dbsp::{
+    algebra::{ZRingValue, ZSet},
+    circuit::{
+        operator_traits::{Data, Operator, SourceOperator},
+        Scope,
+    },
+};
+use rand::Rng;
+use std::{borrow::Cow, marker::PhantomData};
+
+pub struct NexmarkDBSPSource<R: Rng, W, C> {
+    /// The generator for events emitted by this source.
+    // TODO(absoludity): Longer-term, it'd be great to use a client (such as a gRPC client) here to
+    // completely separate the generator and allow the generator to be used with other languages
+    // etc.
+    generator: NexmarkGenerator<R>,
+    batch_size: usize,
+    _t: PhantomData<(C, W)>,
+}
+
+impl<R, W, C> NexmarkDBSPSource<R, W, C>
+where
+    R: Rng,
+{
+    pub fn from_generator(generator: NexmarkGenerator<R>, batch_size: usize) -> Self {
+        NexmarkDBSPSource {
+            generator,
+            batch_size,
+            _t: PhantomData,
+        }
+    }
+}
+
+impl<R, W, C> Operator for NexmarkDBSPSource<R, W, C>
+where
+    C: Data,
+    R: Rng + 'static,
+    W: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("NexmarkDBSPSource")
+    }
+
+    // Do clock_start and clock_end make sense for an infinite source? Or should
+    // the Nexmark generator use the max_events and stop generating events after
+    // that.
+    /// That would also mean then that the fixed point for this data source
+    /// would be at max_events?  Currently the generator will keep spitting out
+    /// events, but maybe `next_event` could be updated to return a
+    /// `Result<Option<NextEvent>>` and return Ok(None) after max events?
+    fn fixedpoint(&self, _scope: Scope) -> bool {
```

Comment:
If the natural behavior for this operator is to generate an infinite sequence of events, then returning `false` is the right thing to do.  If the generator has a predefined limit on the number of outputs, then `eval` should start producing empty sets once the limit has been reached, and `fixedpoint` should return `true` after outputting the first empty set to indicate that all future outputs will be exactly the same.